File: src/pocketcloud/cloud/network/client/ServerClientCache.php
<?php

namespace pocketcloud\cloud\network\client;

use Closure;
use pocketcloud\cloud\server\CloudServer;
use pocketcloud\cloud\server\CloudServerManager;
use pocketcloud\cloud\terminal\log\CloudLogger;
use pocketcloud\cloud\util\net\Address;
use pocketcloud\cloud\util\SingletonTrait;
use pocketcloud\cloud\util\tick\Tickable;

final class ServerClientCache implements Tickable {
    use SingletonTrait;

    /** @var array<ServerClient> */
    private array $clients = [];

    public function __construct() {
        self::setInstance($this);
    }

    public function add(CloudServer $server, ServerClient $client): void {
        if (!$this->isset($client)) {
            CloudLogger::get()->debug("Adding client " . $client->getAddress() . " => " . $server->getName());
            $this->clients[$server->getName()] = $client;
        }
    }

    public function remove(ServerClient|CloudServer $client): void {
        $client = $client instanceof CloudServer ? $this->get($client) : $client;
        if ($client !== null) {
            if ($this->isset($client)) {
                CloudLogger::get()->debug("Removing client " . $client->getAddress());
                unset($this->clients[array_search($client, $this->clients)]);
            }
        }
    }

    public function isset(ServerClient $client): bool {
        return in_array($client, $this->clients);
    }

    public function pick(Closure $conditionClosure): array {
        return array_filter($this->clients, $conditionClosure);
    }

    public function tick(int $currentTick): void {
        $clientsWithDelayedPackets = array_filter($this->clients, fn(ServerClient $client) => count($client->getDelayedPackets()) > 0);
        if (count($clientsWithDelayedPackets) == 0) return;
        foreach ($clientsWithDelayedPackets as $client) {
            foreach ($client->getDelayedPackets() as $i => $data) {
                $packet = $data[0];
                $tick = $data[1];
                if ($tick <= $currentTick) {
                    $sendClosure = $data[2] ?? null;
                    $success = $client->sendPacket($packet);
                    if ($sendClosure !== null) ($sendClosure)($client, $packet, $success);
                    $client->unsetDelayedPacket($i);
                }
            }
        }
    }

    public function get(CloudServer $server): ?ServerClient {
        return $this->clients[$server->getName()] ?? null;
    }

    public function getServer(ServerClient $client): ?CloudServer {
        return $this->isset($client) ? CloudServerManager::getInstance()->get(array_search($client, $this->clients)) : null;
    }

    public function getByAddress(Address $address): ?ServerClient {
        return array_find($this->clients, fn($client) => $client->getAddress()->equals($address));
    }

    public function getAll(): array {
        return $this->clients;
    }
}

File: src/main/java/de/pocketcloud/cloud/network/client/ServerClientCache.java
package de.pocketcloud.cloud.network.client;

import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.tick.Tickable;
import io.netty.channel.Channel;
import lombok.Getter;

import java.net.SocketAddress;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Predicate;

public final class ServerClientCache implements Tickable {

    @Getter
    private static ServerClientCache instance;

    /** serverName -> ServerClient */
    private final Map<String, ServerClient> clientsByServer = new HashMap<>();
    /** client.toString() ->serverName */
    private final Map<String, String> serversByClient = new HashMap<>();
    /** SocketAddress.toString() -> ServerClient */
    private final Map<String, ServerClient> clientsByAddress = new HashMap<>();

    public ServerClientCache() {
        instance = this;
    }

    @Override
    public void tick(long currentTick) {
        for (ServerClient client : clientsByServer.values()) {
            for (ServerClient.DelayedPacket dp : client.pollDuePackets()) {
                boolean success = client.sendPacket((de.pocketcloud.cloud.network.packet.ClientboundPacket) dp.packet());
                if (dp.onSend() != null) {
                    dp.onSend().accept(client, dp.packet(), success);
                }
            }
        }
    }

//    public synchronized void add(CloudServer server, ServerClient client) {
//        if (isRegistered(client)) return;
//        CloudLogger.get().debug("Adding client {} => {}", client, server.getName());
//        clientsByServer.put(server.getName(), client);
//        serversByClient.put(client.toString(), server.getName());
//        clientsByAddress.put(client.getAddress().toString(), client);
//    }

    public synchronized void remove(ServerClient client) {
        if (!isRegistered(client)) return;
        CloudLogger.get().debug("Removing client {}", client);
        String serverName = serversByClient.remove(client.toString());
        if (serverName != null) {
            clientsByServer.remove(serverName);
        } else {
            clientsByServer.values().remove(client);
        }
        clientsByAddress.remove(client.getAddress().toString());
    }

    //TODo
//    public synchronized void remove(CloudServer server) {
//        ServerClient client = clientsByServer.get(server.getName());
//        if (client != null) remove(client);
//    }

    public synchronized boolean isRegistered(ServerClient client) {
        return serversByClient.containsKey(client.toString());
    }

    public synchronized ServerClient getByChannel(Channel channel) {
        return clientsByAddress.get(channel.remoteAddress().toString());
    }

    public synchronized ServerClient getByAddress(SocketAddress address) {
        return clientsByAddress.get(address.toString());
    }

//    public synchronized ServerClient get(CloudServer server) {
//        return clientsByServer.get(server.getName());
//    }
//
//    public synchronized CloudServer getServer(ServerClient client) {
//        String serverName = serversByClient.get(client.toString());
//        if (serverName == null) return null;
//        // Resolve through your CloudServerManager — adjust the call to match your actual API.
//        return de.pocketcloud.cloud.server.CloudServerManager.getInstance().get(serverName);
//    }

    public synchronized List<ServerClient> getAll(Predicate<ServerClient> filter) {
        return clientsByServer.values().stream()
                .filter(filter)
                .toList();
    }

    public synchronized Collection<ServerClient> getAll() {
        return List.copyOf(clientsByServer.values());
    }

    public static long convertTicksToMs(int ticks) {
        return ticks * 50L; // 1000 ms / 20 tps
    }
}
